src/server/search/search_family.cc

```diff
@@ -749,77 +750,129 @@ void SearchFamily::FtSearch(CmdArgList args, Transaction* tx, SinkReplyBuilder*
+    auto shard_start = absl::Now();
+    search_results[shard_id] = index->Search(t->GetOpArgs(es), *params, &search_algo);
+    profile_results[shard_id] = {absl::Now() - shard_start};
 
     return OpStatus::OK;
   });
 
+  if (index_not_found.load())
+    return builder->SendError(std::string{index_name} + ": no such index");
+
   auto took = absl::Now() - start;
+
+  bool result_is_empty = false;
+  size_t total_docs = 0;
+  size_t total_serialized = 0;
+  for (const auto& result : search_results) {
+    if (!result.error) {
+      total_docs += result.total_hits;
+      total_serialized += result.docs.size();
+    } else {
+      result_is_empty = true;
+    }
+  }
+
   auto* rb = static_cast<RedisReplyBuilder*>(builder);
-  rb->StartArray(results.size() + 1);
+  // First element -> Result of the search command
+  // Second element -> Profile information
+  rb->StartArray(2);
+
+  // Result of the search command
+  if (!result_is_empty) {
+    auto agg = search_algo.HasAggregation();
+    if (agg) {
+      ReplySorted(*agg, *params, absl::MakeSpan(search_results), builder);
+    } else {
+      ReplyWithResults(*params, absl::MakeSpan(search_results), builder);
+    }
+  } else {
+    rb->StartArray(1);
+    rb->SendLong(0);
+  }
+
+  // Profile information
+  rb->StartArray(shards_count + 1);
 
   // General stats
   rb->StartCollection(3, RedisReplyBuilder::MAP);
   rb->SendBulkString("took");
   rb->SendLong(absl::ToInt64Microseconds(took));
   rb->SendBulkString("hits");
-  rb->SendLong(total_docs);
+  rb->SendLong(static_cast<long>(total_docs));
   rb->SendBulkString("serialized");
-  rb->SendLong(total_serialized);
+  rb->SendLong(static_cast<long>(total_serialized));
 
   // Per-shard stats
-  for (const auto& [profile, shard_took] : results) {
+  for (size_t shard_id = 0; shard_id < shards_count; shard_id++) {
     rb->StartCollection(2, RedisReplyBuilder::MAP);
     rb->SendBulkString("took");
-    rb->SendLong(absl::ToInt64Microseconds(shard_took));
+    rb->SendLong(absl::ToInt64Microseconds(profile_results[shard_id]));
     rb->SendBulkString("tree");
 
-    for (size_t i = 0; i < profile.events.size(); i++) {
-      const auto& event = profile.events[i];
+    const auto& search_result = search_results[shard_id];
+    if (search_result.error || !search_result.profile || search_result.profile->events.empty()) {
+      rb->SendEmptyArray();
+      continue;
+    }
+
+    const auto& events = search_result.profile->events;
+    for (size_t i = 0; i < events.size(); i++) {
+      const auto& event = events[i];
 
       size_t children = 0;
-      for (size_t j = i + 1; j < profile.events.size(); j++) {
-        if (profile.events[j].depth == event.depth)
+      for (size_t j = i + 1; j < events.size(); j++) {
+        if (events[j].depth == event.depth)
           break;
-        if (profile.events[j].depth == event.depth + 1)
+        if (events[j].depth == event.depth + 1)
```

Comment:
seems weird that at line 886 we have `rb->StartArray(children);` and nothing follows. what does it mean?

Reply:
We are printing tree here. All nodes are just an array. Some events have children. For example we have [Node1, Node2, Node3]. And Node1 is parent of the Node2 and Node3. In this case we will print:
```
1) [Profile info for Node1] // And then we are starting array of size 2 here
2) 1) [Profile info for Node2]
   2) [Profile info for Node3]
```

Reply:
so the next loop will print the profile for children?

Reply:
Yes